Core/Entities : Conversation

Traesh opened the pull request:
**Changes proposed:**
-  Implement Conversation Entity, used to display ingame dialog
-  Command .debug conversation # convId
-  Command .Reload conversation_template
- Add 5 tables for templates, actors & lines

**Target branch(es):** 6.x

**Issues addressed:** Closes #17992 ([Core/Entity : new dialogs in WoW Legion](https://github.com/TrinityCore/TrinityCore/issues/17992))

**Tests performed:** (Does it build, tested in-game, etc)
Built
Tested ingame with conversation id 705    (.debug conv 705)
Tested ingame with conversation id 1025  (.debug conv 1025)

**Known issues and TODO list:**
- None ATM

(Re-Opening PR due to a fail on previous repo, all reviews have been fixed)

---

(minor edit/addendum by tkrokli)

Traesh committed:
Begin working on Conversation entity

Traesh committed:
Finish Conversation System

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into conversation

# Conflicts:
#	src/server/game/Accounts/RBAC.h
#	src/server/game/Globals/ObjectMgr.cpp
#	src/server/game/Globals/ObjectMgr.h
#	src/server/game/World/World.cpp
#	src/server/scripts/Commands/cs_debug.cpp
#	src/server/scripts/Commands/cs_reload.cpp

Traesh committed:
Merge typos

Traesh committed:
Minor improves

Traesh committed:
Fix reviews

Traesh committed:
Typo

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into conversation

Traesh committed:
Rename 9999_99_99_00_auth.sql to 9999_99_99_18010_auth.sql

Traesh committed:
Rename 9999_99_99_00_world.sql to 9999_99_99_18010_world.sql

Pixel-Pirate commented:
Your pr is interesting but I think his is more convenient to make it work with the hotfixes databse

Traesh commented:
Hello :) What do you mean by making it work with the hotfixes database ?

tkrokli commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -0,0 +1,121 @@
+}
+
+
+bool Conversation::CreateConversation(ObjectGuid::LowType guidlow, uint32 conversationEntry, Unit* caster, SpellInfo const* spell, Position const& pos)
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Entities/Conversation/Conversation.cpp:87:125: fatal error: 
      unused parameter 'spell' [-Wunused-parameter]
  ...uint32 conversationEntry, Unit* caster, SpellInfo const* spell, Positio...
                                                              ^
```

tkrokli commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -5681,6 +5682,24 @@ void Spell::EffectCreateGarrison(SpellEffIndex effIndex)
     unitTarget->ToPlayer()->CreateGarrison(GetEffect(effIndex)->MiscValue);
 }
 
+void Spell::EffectCreateConversation(SpellEffIndex effIndex)
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Spells/SpellEffects.cpp:5685:52: fatal error: 
      unused parameter 'effIndex' [-Wunused-parameter]
void Spell::EffectCreateConversation(SpellEffIndex effIndex)
                                                   ^
```

Traesh committed:
Comment unused param (SpellInfo)

Traesh committed:
Comment unused param (Conversation)

tkrokli commented:
Good to see that there is only the 'auth' database left to fix (ToDo when this PR gets merged).
Anything else left to check or review in this PR? :)

Druson commented:
seems ConversationLine.db2 contain the next conversation  line ID of current conversation line.  So can table conversation_lines just contain the conversation ID and first line ID only? 

And the field Unk1 in the table conversation_line_template should be the duration of previous conversation_line. The duration should be different for different locale.

funjoker commented:
News?

Aokromes commented:
continuous-integration/travis-ci/pr — The Travis CI build failed

funjoker commented:
Travis fail due to SQL?

tkrokli commented:
Yes, it is only because the auth database update list is not adjusted yet.
```
- "sql/updates/auth/9999_99_99_18010_auth.sql" is missing in the 'auth'.'updates' table.
Fatal error:

  The Database Updater is broken for the 'auth' database,
  because the applied update is missing in the 'auth'.'updates' table.

How to fix:
  Insert the missing names of the already applied sql updates
  to the 'updates' table of the 'auth' base dump ('sql/base/auth_database.sql').
```
Then again, even if the update is added in the PR, it needs to be redone after merge.
So it won't make much difference if it is included in the PR, mostly only to the Travis result.

funjoker commented:
Merge?

tkrokli commented:
This merge needs to be done by someone with experience in creating the correct hash change to `auth`.`updates` in addition to the normal renaming of the SQL files in this PR.

Naios commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -5681,6 +5682,24 @@ void Spell::EffectCreateGarrison(SpellEffIndex effIndex)
+
+    uint32 conversationEntry = effectInfo->MiscValue;
+
+    Conversation* conversation = new Conversation;
```
Use snart pointers please

Shauren replied:
and noooo

Naios commented on src/server/game/Globals/ObjectMgr.h:
```diff
@@ -1424,6 +1430,15 @@ class TC_GAME_API ObjectMgr
+            if (itr != _conversationTemplateStore.end())
+                return &itr->second;
+
+            return nullptr;
```
You could use `Optional` here to clarify that your pointer might be null.

Shauren replied:
noooo

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -0,0 +1,121 @@
+{
+    if (IsInWorld())
+    {
+        SendObjectDeSpawnAnim(GetGUID());
```
it's a gameobject only packet

joschiwald commented:
what is the current state of this pr?

funjoker commented:
well he is active thing he finishes areatrigger first

Traesh committed:
Merge branch 'master' of github.com:TrinityCore/TrinityCore into conversation

Traesh committed:
Fix reviews

tkrokli commented:
BTW, in your next commit, please make sure that you change the location of your auth sql update:
`sql/updates/auth/6.x/9999_99_99_18010_auth.sql` needs to be moved to
`sql/updates/auth/master/9999_99_99_18010_auth.sql`

matukaa commented:
Umm, up? :D

Commit:
Merge branch 'master' into conversation

tkrokli commented:
Going to push another commit in this PR branch, moving the `[ ]_auth.sql` from `6.x` to `master`.

joschiwald commented on sql/updates/auth/6.x/9999_99_99_18010_auth.sql:
```diff
@@ -0,0 +1,7 @@
+DELETE FROM `rbac_permissions` WHERE `id` IN (851, 852);
+INSERT INTO `rbac_permissions` (`id`, `name`) VALUES
+(851, 'Command: .reload conversation_template'),
+(852, 'Command: .debug conversation');
```
wrong ids

tkrokli replied:
Thanks, I will fix that too. Forgot to change it outside of solving the conflict.

Commit:
Update auth sql file to TC master branch

- renamed & moved 9999_99_99_18010_auth.sql
  to sql/updates/auth/master/2017_99_99_00_auth.sql
- changed RBAC permission numbers to match the .cpp change
- corrected the table name used in HotfixDatabaseConnection::DoPrepareStatements
- corrected the table name used in 2017_99_99_18010_hotfixes.sql
- added the auth DB update with temporary name 2017_99_99_00_auth.sql

tkrokli commented on sql/updates/hotfixes/master/9999_99_99_18010_hotfixes.sql:
```diff
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS `conversation_line`;
```
This `DROP TABLE` command does not match `CREATE TABLE`.
Going to remedy this in my commit as well.

tkrokli commented on src/server/database/Database/Implementation/HotfixDatabase.cpp:
```diff
@@ -175,6 +175,10 @@ void HotfixDatabaseConnection::DoPrepareStatements()
 
+    // ConversationLine.db2
+    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
+        " FROM conversation_line ORDER BY ID DESC", CONNECTION_SYNCH);
```
```
In mysql_stmt_prepare() id: 48, sql: "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6 FROM conversation_line ORDER BY ID DESC"
Table 'hotfixes.conversation_line' doesn't exist
```
I presume the intended table name is `conversation_line_template`.
Adding this fix to my commit too.

tkrokli replied:
Sorry, I was wrong. The correct table name here is `conversation_template`.
Fix rework still in progress.

joschiwald commented:
Don't touch hotfix database

tkrokli commented:
I only renamed the table name `conversation_line` to `conversation_template`, which was already referenced in the already existing files `sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql` and `src/server/database/Database/Implementation/HotfixDatabase.cpp`. Maybe the error message I posted yesterday was a bit confusing or misleading?

Shauren commented:
Thats where you are wrong. hotfix tables are autogenerated and must strictly follow naming of its corresponding .db2 file

Commit:
partial revert of commit 9092546

Removing changes related to Hotfixes

tkrokli commented:
As you wish, reverting my changes related to Hotfixes.

tkrokli commented:
I don't get it. The PR compiles, but my local worldserver won't complete starting up. From Server.log:
```
Could not prepare statements of the Hotfix database, see log for details.
```
From DBErrors.log:
```
In mysql_stmt_prepare() id: 48, sql: "SELECT ID, Unk1, Unk2, Unk3, NextLineID, Unk4, UnkBoolean, Unk5, Unk6 FROM conversation_line ORDER BY ID DESC"
Table 'hotfixes.conversation_line' doesn't exist
```
This happens while starting worldserver, rebuilding all the DBs from scratch (removed all DBs before start).

Shauren commented:
Look closely at that sql file
```
DROP TABLE IF EXISTS `conversation_line`;
CREATE TABLE `conversation_template` (
```

tkrokli commented:
Sure, that is one of the 2 things I corrected in my previous commit, but I was asked to revert it.
Maybe I am too slow (or braindead) to see what the correct solution is. Please enlighten me.

Shauren commented:
Change the second line only to be `conversation_line`, not the first line and core

Commit:
Update 2017_99_99_18010_hotfixes.sql

Set `conversation_line` as the correct name for the new table.

tkrokli commented:
Any other changes required in this PR before merging?

joschiwald commented on sql/updates/hotfixes/master/2017_99_99_18010_hotfixes.sql:
```diff
@@ -0,0 +1,14 @@
+  `Unk4` smallint (5) UNSIGNED NOT NULL,
+  `UnkBoolean` tinyint (3) UNSIGNED NOT NULL,
+  `Unk5` tinyint (3) UNSIGNED NOT NULL,
+  `Unk6` tinyint (3) UNSIGNED NOT NULL,
```
one of these unknowns is a string field, but the dbcviewer does not show any useful things

joschiwald replied:
@Shauren: any idea what are these unknowns?

funjoker replied:
no don't think there is a string inside

funjoker replied:
Nope no strings.
All strings are inside the broadcast_text at hotfixes db.
Unk1 is the broadcast_text.Id !

funjoker replied:
Unk3 is a break between the running and following Text in milliseconds

joschiwald commented on src/server/game/Globals/ObjectMgr.cpp:
```diff
@@ -9570,3 +9570,108 @@ void ObjectMgr::LoadSceneTemplates()
+                {
+                    Field* fields = actors->Fetch();
+                    uint32 actorId = fields[0].GetUInt32();
+                    ConversationActorTemplate conversationActorTemplate = _conversationActorTemplateStore[actorId];
```
ConversationActorTemplate&

joschiwald commented on src/server/game/Globals/ObjectMgr.cpp:
```diff
@@ -9570,3 +9570,108 @@ void ObjectMgr::LoadSceneTemplates()
+
+            while (currentConversationLine != nullptr)
+            {
+                ConversationLineTemplate conversationActorTemplate = _conversationLineTemplateStore[currentConversationLine->ID];
```
ref

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -0,0 +1,123 @@
+    _duration = conversationTemplate->LastLineDuration;
+
+    uint16 actorsIndex = 0;
+    for (ConversationActorTemplate actor : conversationTemplate->Actors)
```
const&

joschiwald commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -0,0 +1,123 @@
+        SetDynamicStructuredValue(CONVERSATION_DYNAMIC_FIELD_ACTORS, actorsIndex++, &actor);
+
+    uint16 linesIndex = 0;
+    for (ConversationLineTemplate line : conversationTemplate->Lines)
```
const&

Commit:
Add missing  refs + const& in 2 files

- Conversation.cpp
- ObjectMgr.cpp

funjoker commented:
Any news about it ? :D

funjoker commented:
```
DROP TABLE IF EXISTS `conversation_line`;
CREATE TABLE `conversation_line` (
  `Id` int (10) UNSIGNED NOT NULL,
  `Unk1` int (10) UNSIGNED NOT NULL,
  `Unk2` int (10) UNSIGNED NOT NULL,
  `Unk3` int (10) UNSIGNED NOT NULL,
  `NextLineID` smallint (5) UNSIGNED NOT NULL,
  `Unk4` smallint (5) UNSIGNED NOT NULL,
  `UnkBoolean` tinyint (3) UNSIGNED NOT NULL,
  `Unk5` tinyint (3) UNSIGNED NOT NULL,
  `Unk6` tinyint (3) UNSIGNED NOT NULL,
  `VerifiedBuild` int (10) UNSIGNED NOT NULL DEFAULT 0,
  PRIMARY KEY (`Id`)
) ENGINE=MyISAM DEFAULT CHARSET=utf8;
```

Unk1 is the broadcast_text.Id
So is the db2 too

funjoker commented:
Unk3 is a break between the running and following Text in milliseconds

tkrokli commented:
OK, so maybe we could borrow the same naming style as for the `broadcast_text.ID` in the `creature_text` table, which is `BroadcastTextId` as a name for the column `Unk1`? (Just a suggestion.)

I am more uncertain about how to pick a good name to replace `Unk3`, because I don't remember an existing column using that kind of information (but I may have overlooked something). I presume the name could usefully contain a 3-part naming reference to
1. conversation or text
2. break or interval
3. ms (milliseconds) or time

Some possible suggestions could be `text_interval_ms` or `ConversationBreakTime`. Maybe there are naming conventions to be followed, I don't remember at the moment. Not sure if it is practical to add that change in this PR, though.

funjoker commented:
I think ConversationBreakTime might be bit long but good.

tkrokli commented:
OK, it was just a suggestion. Shorter suggestions could be `TextBreakMS` / `LineIntervalMS` and so on.

offl commented:
```sql
  `NextLineTime` int (10) UNSIGNED NOT NULL,
  `NextLineID` smallint (5) UNSIGNED NOT NULL,
```

funjoker commented:
> ```
>  `NextLineTime` int (10) UNSIGNED NOT NULL,
>   `NextLineID` smallint (5) UNSIGNED NOT NULL,
> ```

Could maybe be confusing about the thing the column is for

Traesh committed:
Merge branch 'master' of github.com:TrinityCore/TrinityCore into conversation

Traesh committed:
Fix SQL actors

Traesh committed:
Define some Unks

funjoker commented:
One of the Unk in conversation_line is the type of the conversation, but donno which =/

funjoker commented:
```
 struct ConversationLineEntry
 {
     uint32 ID;
     uint32 BroadcastTextId;
     uint32 Unk2;
     uint32 LineIntervalMS;
     uint16 NextLineID;
     uint16 Unk4; // Flag ?
     uint8 UnkBoolean;
     uint8 Unk5;
     uint8 Unk6;
 };
```

Unk2; <--- SpellVisualKitID

funjoker commented:
```
struct ConversationLineTemplate
 {
     uint32 Id;      // Link to ConversationLine.db2
     uint32 PreviousLineDuration;
     uint32 Unk2;
     uint32 Unk3;
 };
```

`Unk2` is it maybe AnimKitId ?
Because the avatars are sometimes using angry movement.

funjoker commented:
SPELL_EFFECT_219  -> SPELL_EFFECT_CONVERSATION 
Not proofed :D at work atm so someone might take a look ^^

Traesh commented:
@funjoker It's already in this PR :p

funjoker commented:
OKAY :D sorry ^^

Aokromes commented:
Conflicting files
sql/base/auth_database.sql
src/server/game/Accounts/RBAC.h

Commit:
Merge branch 'master' into conversation

tkrokli commented on src/server/game/DataStores/DB2Stores.cpp:
```diff
@@ -58,6 +58,7 @@ DB2Storage<ChrRacesEntry>                       sChrRacesStore("ChrRaces.db2", C
 DB2Storage<ChrSpecializationEntry>              sChrSpecializationStore("ChrSpecialization.db2", ChrSpecializationLoadInfo::Instance());
 DB2Storage<CinematicCameraEntry>                sCinematicCameraStore("CinematicCamera.db2", CinematicCameraLoadInfo::Instance());
 DB2Storage<CinematicSequencesEntry>             sCinematicSequencesStore("CinematicSequences.db2", CinematicSequencesLoadInfo::Instance());
+DB2Storage<ConversationLineEntry>               sConversationLineStore("ConversationLine.db2", ConversationLineLoadInfo::Instance());
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/DataStores/DB2Stores.cpp:61:49: fatal error: 
      no matching constructor for initialization of
      'DB2Storage<ConversationLineEntry>'
  ...sConversationLineStore("ConversationLine.db2", ConversationLineLoadInfo::Instance());
     ^                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Shauren replied:
Yeah, I changed the loadinfo part when making db2 loaders usable by extractors, PR author needs to rerun latest https://github.com/Shauren/WowClientDB2MySQLTableGenerator

Traesh committed:
Fix ConversationLineLoadInfo

Traesh committed:
Typo

wonder2011 commented:
Merge?

Pixel-Pirate commented on src/server/database/Database/Implementation/HotfixDatabase.cpp:
```diff
@@ -180,7 +180,7 @@ void HotfixDatabaseConnection::DoPrepareStatements()
 
     // ConversationLine.db2
-    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, BroadcastTextId, Unk2, LineIntervalMS, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
+    PrepareStatement(HOTFIX_SEL_CONVERSATION_LINE, "SELECT ID, BroadcastTextId, BroadcastTextId, LineIntervalMS, NextLineID, Unk4, UnkBoolean, Unk5, Unk6"
```
Not BroadcastTextId but SpellVisualKitID ;)

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into conversation

# Conflicts:
#	sql/base/auth_database.sql

Traesh committed:
Fix typo

Traesh committed:
Fix for 7.2

Shauren commented on src/server/game/Entities/Conversation/Conversation.cpp:
```diff
@@ -102,7 +102,7 @@ bool Conversation::CreateConversation(ObjectGuid::LowType guidlow, uint32 conver
         return false;
 
-    SetUInt32Value(CONVERSATION_FIELD_LAST_LINE_DURATION, conversationTemplate->LastLineDuration);
+    SetUInt32Value(CONVERSATION_LAST_LINE_END_TIME, conversationTemplate->LastLineDuration);
```
well, are you sure about this now?

joschiwald commented on src/server/game/Entities/Conversation/Conversation.h:
```diff
@@ -0,0 +1,76 @@
+
+struct ConversationActorTemplate
+{
+    uint32 Id;
```
the first 4 uint can be a full guid

joschiwald committed:
Added ConversationDataStore

joschiwald committed:
Merge branch 'master' of github.com:TrinityCore/TrinityCore into conversation

joschiwald committed:
Named some unknowns

Shauren commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -0,0 +1,170 @@
+                    Field* fields = actors->Fetch();
+                    uint32 actorId = fields[0].GetUInt32();
+                    if (ConversationActorTemplate const* conversationActorTemplate = Trinity::Containers::MapGetValuePtr(_conversationActorTemplateStore, actorId))
+                        conversationTemplate.Actors.push_back(*conversationActorTemplate);
```
Whats the logic behind just copying the actor inside the conversation template?

joschiwald replied:
i don't see it too, i just copied stuff from ObjectMgr with few modifications, i'm going to change something about saving actors after i have move infos about them, do you have some? about the unknowns?

joschiwald committed:
Named some unknowns in actor template

joschiwald committed:
Conversations are only visible to specific players

Shauren commented on src/server/game/Globals/ConversationDataStore.cpp:
```diff
@@ -0,0 +1,179 @@
+
+#include "ConversationDataStore.h"
+#include "Containers.h"
+#include "Conversation.h"
```
This goes against my idea of cleaning up headers - it should be `ConversationDataStore.h` that defines the template structures (or even a separate header), not `Conversation.h` (heavy header to include due to Object.h - pulls all of map & grid)

joschiwald committed:
Moved template definition to ConversationDataStore

joschiwald committed:
Fixed build

joschiwald commented:
i'm going to merge this pr tomorrow

joschiwald committed:
fixed build

joschiwald committed:
Prepare merge
